vhdl_lang/src/syntax/separated_list.rs

```diff
@@ -5,39 +5,71 @@
 // Copyright (c) 2023, Olof Kraigher olof.kraigher@gmail.com
 
 use crate::ast::{IdentList, NameList, SeparatedList, WithRef};
-use crate::data::DiagnosticResult;
+use crate::data::{DiagnosticHandler, DiagnosticResult};
 use crate::syntax::common::ParseResult;
 use crate::syntax::names::parse_name;
 use crate::syntax::Kind::Comma;
-use crate::syntax::{Kind, TokenStream};
+use crate::syntax::{kind_str, Kind, TokenAccess, TokenStream};
+use std::fmt::Debug;
+
+/// Skip extraneous tokens of kind `separator`.
+/// When there are any extra tokens of that kind, mark all the positions of these tokens as erroneous
+fn skip_extraneous_tokens(
+    stream: &TokenStream,
+    separator: Kind,
+    diagnostics: &mut dyn DiagnosticHandler,
+) {
+    if let Some(separator_tok) = stream.pop_if_kind(separator) {
+        let start_pos = stream.get_pos(separator_tok);
+        let mut end_pos = start_pos;
+        while let Some(separator_tok) = stream.pop_if_kind(separator) {
+            end_pos = stream.get_pos(separator_tok)
+        }
+        diagnostics.error(
+            start_pos.combine(end_pos),
+            format!("Extraneous '{}'", kind_str(separator)),
+        );
+    }
+}
 
 /// Parses a list of the form
 ///   `element { separator element }`
 /// where `element` is an AST element and `separator` is a token of some `ast::Kind`.
 /// The returned list retains information of the whereabouts of the separator tokens.
-pub fn parse_list_with_separator<F, T>(
+pub fn parse_list_with_separator<F, T: Debug>(
     stream: &TokenStream,
     separator: Kind,
+    diagnostics: &mut dyn DiagnosticHandler,
     parse_fn: F,
 ) -> DiagnosticResult<SeparatedList<T>>
 where
     F: Fn(&TokenStream) -> ParseResult<T>,
 {
     let mut items = vec![parse_fn(stream)?];
     let mut tokens = Vec::new();
-    while let Some(separator) = stream.pop_if_kind(separator) {
-        items.push(parse_fn(stream)?);
-        tokens.push(separator);
+    while let Some(separator_tok) = stream.pop_if_kind(separator) {
+        skip_extraneous_tokens(stream, separator, diagnostics);
+        tokens.push(separator_tok);
+        match parse_fn(stream) {
```

Comment:
The first call to `parse_fun` uses the ?-operator for early exit. But inside the loop there is no longer early exit for `parse_fun` failures after this change. This is a bit inconsistent.

I do not see any test case that requires the inner loop `parse_fn` to not be early exit. Are you sure it is needed and was not an early idea that was then later not used?

Reply:
This is indeed inconsistent. Do you think my change (not short-circuiting when the inner `parse_fn` fails) or the former (always short-circuiting) is better?
Personally, I think that the new change might improve completion (i.e. `a => ,` will be incorrect but completions still work). However, I'm not sure if this leads to spurious error messages. What do you think?

Reply:
I think it is ok as long as there is a test for the case where it provides value.